// carbon-components-react/lib/components/ComposedModal/ComposedModal-story.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WithStateManager = exports.PassiveModal = exports.Default = exports.Playground = exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactDom = _interopRequireDefault(require("react-dom"));

var _addonActions = require("@storybook/addon-actions");

var _addonKnobs = require("@storybook/addon-knobs");

var _ComposedModal = _interopRequireWildcard(require("../ComposedModal"));

var _Select = _interopRequireDefault(require("../Select"));

var _SelectItem = _interopRequireDefault(require("../SelectItem"));

var _TextInput = _interopRequireDefault(require("../TextInput"));

var _Button = _interopRequireDefault(require("../Button"));

var _ComposedModal2 = _interopRequireDefault(require("./ComposedModal.mdx"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr && (typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]); if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var sizes = {
  'Extra small (xs)': 'xs',
  'Small (sm)': 'sm',
  'Medium (md)': 'md',
  'Large (lg)': 'lg'
};
var buttons = {
  'None (0)': '0',
  'One (1)': '1',
  'Two (2)': '2',
  'Three (3)': '3'
};
var props = {
  composedModal: function composedModal() {
    return {
      numberOfButtons: (0, _addonKnobs.optionsKnob)('Number of Buttons', buttons, '2', {
        display: 'inline-radio'
      }),
      open: (0, _addonKnobs.boolean)('Open (open in <ComposedModal>)', true),
      onKeyDown: (0, _addonActions.action)('onKeyDown'),
      selectorPrimaryFocus: (0, _addonKnobs.text)('Primary focus element selector (selectorPrimaryFocus)', '[data-modal-primary-focus]'),
      size: (0, _addonKnobs.select)('Size (size)', sizes, 'md'),
      preventCloseOnClickOutside: (0, _addonKnobs.boolean)('Prevent closing on click outside of modal (preventCloseOnClickOutside)', true)
    };
  },
  modalHeader: function modalHeader() {
    var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
        titleOnly = _ref.titleOnly;

    return {
      label: (0, _addonKnobs.text)('Optional Label (label in <ModalHeader>)', 'Label'),
      title: (0, _addonKnobs.text)('Optional title (title in <ModalHeader>)', titleOnly ? "\n      Passive modal title as the message. Should be direct and 3 lines or less.\n    ".trim() : 'Modal heading'),
      iconDescription: (0, _addonKnobs.text)('Close icon description (iconDescription in <ModalHeader>)', 'Close'),
      buttonOnClick: (0, _addonActions.action)('buttonOnClick')
    };
  },
  modalBody: function modalBody() {
    return {
      hasScrollingContent: (0, _addonKnobs.boolean)('Modal contains scrollable content (hasScrollingContent)', false),
      'aria-label': (0, _addonKnobs.text)('ARIA label for content', 'Example modal content')
    };
  },
  modalFooter: function modalFooter(numberOfButtons) {
    var secondaryButtons = function secondaryButtons() {
      switch (numberOfButtons) {
        case '2':
          return {
            secondaryButtonText: (0, _addonKnobs.text)('Secondary button text (secondaryButtonText in <ModalFooter>)', 'Secondary button')
          };

        case '3':
          return {
            secondaryButtons: (0, _addonKnobs.object)('Secondary button config array (secondaryButtons)', [{
              buttonText: 'Keep both',
              onClick: (0, _addonActions.action)('onClick')
            }, {
              buttonText: 'Rename',
              onClick: (0, _addonActions.action)('onClick')
            }])
          };

        default:
          return null;
      }
    };

    return _objectSpread(_objectSpread({
      danger: (0, _addonKnobs.boolean)('Primary button danger (danger)', false),
      primaryButtonText: (0, _addonKnobs.text)('Primary button text (primaryButtonText in <ModalFooter>)', 'Primary button'),
      primaryButtonDisabled: (0, _addonKnobs.boolean)('Primary button disabled (primaryButtonDisabled in <ModalFooter>)', false)
    }, secondaryButtons(numberOfButtons)), {}, {
      onRequestClose: (0, _addonActions.action)('onRequestClose'),
      onRequestSubmit: (0, _addonActions.action)('onRequestSubmit')
    });
  }
};

var scrollingContent = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("p", null, "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aenean id accumsan augue. Phasellus consequat augue vitae tellus tincidunt posuere. Curabitur justo urna, consectetur vel elit iaculis, ultrices condimentum risus. Nulla facilisi. Etiam venenatis molestie tellus. Quisque consectetur non risus eu rutrum.", ' '), /*#__PURE__*/_react.default.createElement("p", null, "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aenean id accumsan augue. Phasellus consequat augue vitae tellus tincidunt posuere. Curabitur justo urna, consectetur vel elit iaculis, ultrices condimentum risus. Nulla facilisi. Etiam venenatis molestie tellus. Quisque consectetur non risus eu rutrum.", ' '), /*#__PURE__*/_react.default.createElement("p", null, "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aenean id accumsan augue. Phasellus consequat augue vitae tellus tincidunt posuere. Curabitur justo urna, consectetur vel elit iaculis, ultrices condimentum risus. Nulla facilisi. Etiam venenatis molestie tellus. Quisque consectetur non risus eu rutrum.", ' '), /*#__PURE__*/_react.default.createElement("h3", null, "Lorem ipsum"), /*#__PURE__*/_react.default.createElement("p", null, "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aenean id accumsan augue. Phasellus consequat augue vitae tellus tincidunt posuere. Curabitur justo urna, consectetur vel elit iaculis, ultrices condimentum risus. Nulla facilisi. Etiam venenatis molestie tellus. Quisque consectetur non risus eu rutrum.", ' '), /*#__PURE__*/_react.default.createElement("p", null, "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aenean id accumsan augue. Phasellus consequat augue vitae tellus tincidunt posuere. Curabitur justo urna, consectetur vel elit iaculis, ultrices condimentum risus. Nulla facilisi. Etiam venenatis molestie tellus. Quisque consectetur non risus eu rutrum.", ' '), /*#__PURE__*/_react.default.createElement("p", null, "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aenean id accumsan augue. Phasellus consequat augue vitae tellus tincidunt posuere. Curabitur justo urna, consectetur vel elit iaculis, ultrices condimentum risus. Nulla facilisi. Etiam venenatis molestie tellus. Quisque consectetur non risus eu rutrum.", ' '), /*#__PURE__*/_react.default.createElement("p", null, "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aenean id accumsan augue. Phasellus consequat augue vitae tellus tincidunt posuere. Curabitur justo urna, consectetur vel elit iaculis, ultrices condimentum risus. Nulla facilisi. Etiam venenatis molestie tellus. Quisque consectetur non risus eu rutrum.", ' '));

var _default = {
  title: 'Components/ComposedModal',
  decorators: [_addonKnobs.withKnobs],
  parameters: {
    component: _ComposedModal.default,
    subcomponents: {
      ModalHeader: _ComposedModal.ModalHeader,
      ModalBody: _ComposedModal.ModalBody,
      ModalFooter: _ComposedModal.ModalFooter
    },
    docs: {
      page: _ComposedModal2.default
    }
  }
};
exports.default = _default;

var Playground = function Playground() {
  var _props$composedModal = props.composedModal(),
      size = _props$composedModal.size,
      numberOfButtons = _props$composedModal.numberOfButtons,
      rest = _objectWithoutProperties(_props$composedModal, ["size", "numberOfButtons"]);

  var _props$modalBody = props.modalBody(),
      hasScrollingContent = _props$modalBody.hasScrollingContent;

  return /*#__PURE__*/_react.default.createElement(_ComposedModal.default, _extends({}, rest, {
    size: size || undefined
  }), /*#__PURE__*/_react.default.createElement(_ComposedModal.ModalHeader, props.modalHeader()), /*#__PURE__*/_react.default.createElement(_ComposedModal.ModalBody, _extends({}, props.modalBody(), {
    "aria-label": hasScrollingContent ? 'Modal content' : undefined
  }), /*#__PURE__*/_react.default.createElement("p", {
    style: {
      marginBottom: '1rem'
    }
  }, "Custom domains direct requests for your apps in this Cloud Foundry organization to a URL that you own. A custom domain can be a shared domain, a shared subdomain, or a shared domain and host."), /*#__PURE__*/_react.default.createElement(_TextInput.default, {
    "data-modal-primary-focus": true,
    id: "text-input-1",
    labelText: "Domain name",
    placeholder: "e.g. github.com",
    style: {
      marginBottom: '1rem'
    }
  }), /*#__PURE__*/_react.default.createElement(_Select.default, {
    id: "select-1",
    defaultValue: "us-south",
    labelText: "Region"
  }, /*#__PURE__*/_react.default.createElement(_SelectItem.default, {
    value: "us-south",
    text: "US South"
  }), /*#__PURE__*/_react.default.createElement(_SelectItem.default, {
    value: "us-east",
    text: "US East"
  })), /*#__PURE__*/_react.default.createElement("br", null), hasScrollingContent && scrollingContent), numberOfButtons > 0 && /*#__PURE__*/_react.default.createElement(_ComposedModal.ModalFooter, props.modalFooter(numberOfButtons)));
};

exports.Playground = Playground;

var Default = function Default() {
  return /*#__PURE__*/_react.default.createElement(_ComposedModal.default, {
    open: true
  }, /*#__PURE__*/_react.default.createElement(_ComposedModal.ModalHeader, {
    label: "Account resources",
    title: "Add a custom domain"
  }), /*#__PURE__*/_react.default.createElement(_ComposedModal.ModalBody, null, /*#__PURE__*/_react.default.createElement("p", {
    style: {
      marginBottom: '1rem'
    }
  }, "Custom domains direct requests for your apps in this Cloud Foundry organization to a URL that you own. A custom domain can be a shared domain, a shared subdomain, or a shared domain and host."), /*#__PURE__*/_react.default.createElement(_TextInput.default, {
    "data-modal-primary-focus": true,
    id: "text-input-1",
    labelText: "Domain name",
    placeholder: "e.g. github.com",
    style: {
      marginBottom: '1rem'
    }
  }), /*#__PURE__*/_react.default.createElement(_Select.default, {
    id: "select-1",
    defaultValue: "us-south",
    labelText: "Region"
  }, /*#__PURE__*/_react.default.createElement(_SelectItem.default, {
    value: "us-south",
    text: "US South"
  }), /*#__PURE__*/_react.default.createElement(_SelectItem.default, {
    value: "us-east",
    text: "US East"
  }))), /*#__PURE__*/_react.default.createElement(_ComposedModal.ModalFooter, {
    primaryButtonText: "Add",
    secondaryButtonText: "Cancel"
  }));
};

exports.Default = Default;
Default.story = {
  name: 'Composed Modal'
};

var PassiveModal = function PassiveModal() {
  return /*#__PURE__*/_react.default.createElement(_ComposedModal.default, {
    open: true
  }, /*#__PURE__*/_react.default.createElement(_ComposedModal.ModalHeader, {
    title: "You have been successfully signed out"
  }), /*#__PURE__*/_react.default.createElement(_ComposedModal.ModalBody, null));
};

exports.PassiveModal = PassiveModal;

var WithStateManager = function WithStateManager() {
  /**
   * Simple state manager for modals.
   */
  var ModalStateManager = function ModalStateManager(_ref2) {
    var LauncherContent = _ref2.renderLauncher,
        ModalContent = _ref2.children;

    var _useState = (0, _react.useState)(false),
        _useState2 = _slicedToArray(_useState, 2),
        open = _useState2[0],
        setOpen = _useState2[1];

    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, !ModalContent || typeof document === 'undefined' ? null : /*#__PURE__*/_reactDom.default.createPortal( /*#__PURE__*/_react.default.createElement(ModalContent, {
      open: open,
      setOpen: setOpen
    }), document.body), LauncherContent && /*#__PURE__*/_react.default.createElement(LauncherContent, {
      open: open,
      setOpen: setOpen
    }));
  };

  return /*#__PURE__*/_react.default.createElement(ModalStateManager, {
    renderLauncher: function renderLauncher(_ref3) {
      var setOpen = _ref3.setOpen;
      return /*#__PURE__*/_react.default.createElement(_Button.default, {
        onClick: function onClick() {
          return setOpen(true);
        }
      }, "Launch composed modal");
    }
  }, function (_ref4) {
    var open = _ref4.open,
        setOpen = _ref4.setOpen;
    return /*#__PURE__*/_react.default.createElement(_ComposedModal.default, {
      open: open,
      onClose: function onClose() {
        return setOpen(false);
      }
    }, /*#__PURE__*/_react.default.createElement(_ComposedModal.ModalHeader, {
      label: "Account resources",
      title: "Add a custom domain"
    }), /*#__PURE__*/_react.default.createElement(_ComposedModal.ModalBody, null, /*#__PURE__*/_react.default.createElement("p", {
      style: {
        marginBottom: '1rem'
      }
    }, "Custom domains direct requests for your apps in this Cloud Foundry organization to a URL that you own. A custom domain can be a shared domain, a shared subdomain, or a shared domain and host."), /*#__PURE__*/_react.default.createElement(_TextInput.default, {
      "data-modal-primary-focus": true,
      id: "text-input-1",
      labelText: "Domain name",
      placeholder: "e.g. github.com",
      style: {
        marginBottom: '1rem'
      }
    }), /*#__PURE__*/_react.default.createElement(_Select.default, {
      id: "select-1",
      defaultValue: "us-south",
      labelText: "Region"
    }, /*#__PURE__*/_react.default.createElement(_SelectItem.default, {
      value: "us-south",
      text: "US South"
    }), /*#__PURE__*/_react.default.createElement(_SelectItem.default, {
      value: "us-east",
      text: "US East"
    }))), /*#__PURE__*/_react.default.createElement(_ComposedModal.ModalFooter, {
      primaryButtonText: "Add",
      secondaryButtonText: "Cancel"
    }));
  });
};

exports.WithStateManager = WithStateManager;
