// carbon-components-react/lib/components/ContextMenu/ContextMenu.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _classnames2 = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _carbonComponents = require("carbon-components");

var _keyboard = require("../../internal/keyboard");

var _ClickListener = _interopRequireDefault(require("../../internal/ClickListener"));

var _utils = require("./_utils");

var _ContextMenuGroup = _interopRequireDefault(require("./ContextMenuGroup"));

var _ContextMenuRadioGroup = _interopRequireDefault(require("./ContextMenuRadioGroup"));

var _ContextMenuRadioGroupOptions = _interopRequireDefault(require("./ContextMenuRadioGroupOptions"));

var _ContextMenuSelectableItem = _interopRequireDefault(require("./ContextMenuSelectableItem"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr && (typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]); if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var prefix = _carbonComponents.settings.prefix;
var margin = 16; // distance to keep to body edges, in px

var ContextMenu = function ContextMenu(_ref) {
  var _classnames;

  var children = _ref.children,
      open = _ref.open,
      _ref$level = _ref.level,
      level = _ref$level === void 0 ? 1 : _ref$level,
      _ref$x = _ref.x,
      x = _ref$x === void 0 ? 0 : _ref$x,
      _ref$y = _ref.y,
      y = _ref$y === void 0 ? 0 : _ref$y,
      _ref$onClose = _ref.onClose,
      onClose = _ref$onClose === void 0 ? function () {} : _ref$onClose,
      rest = _objectWithoutProperties(_ref, ["children", "open", "level", "x", "y", "onClose"]);

  var rootRef = (0, _react.useRef)(null);

  var _useState = (0, _react.useState)(1),
      _useState2 = _slicedToArray(_useState, 2),
      direction = _useState2[0],
      setDirection = _useState2[1]; // 1 = to right, -1 = to left


  var _useState3 = (0, _react.useState)([x, y]),
      _useState4 = _slicedToArray(_useState3, 2),
      position = _useState4[0],
      setPosition = _useState4[1];

  var _useState5 = (0, _react.useState)(false),
      _useState6 = _slicedToArray(_useState5, 2),
      canBeClosed = _useState6[0],
      setCanBeClosed = _useState6[1];

  var isRootMenu = level === 1;

  function focusNode(node) {
    if (node) {
      var _rootRef$current;

      (0, _utils.resetFocus)(rootRef === null || rootRef === void 0 ? void 0 : (_rootRef$current = rootRef.current) === null || _rootRef$current === void 0 ? void 0 : _rootRef$current.element);
      (0, _utils.focusNode)(node);
    }
  }

  function handleKeyDown(event) {
    if (event.target.tagName === 'LI' && ((0, _keyboard.match)(event, _keyboard.keys.Enter) || (0, _keyboard.match)(event, _keyboard.keys.Space))) {
      handleClick(event);
    } else {
      event.stopPropagation();
    }

    if ((0, _keyboard.match)(event, _keyboard.keys.Escape) || !isRootMenu && (0, _keyboard.match)(event, _keyboard.keys.ArrowLeft)) {
      onClose();
    }

    var nodeToFocus;

    if (event.target.tagName === 'LI') {
      var currentNode = event.target;

      if ((0, _keyboard.match)(event, _keyboard.keys.ArrowUp)) {
        nodeToFocus = (0, _utils.getNextNode)(currentNode, -1);
      } else if ((0, _keyboard.match)(event, _keyboard.keys.ArrowDown)) {
        nodeToFocus = (0, _utils.getNextNode)(currentNode, 1);
      } else if ((0, _keyboard.match)(event, _keyboard.keys.ArrowLeft)) {
        nodeToFocus = (0, _utils.getParentNode)(currentNode);
      }
    } else if (event.target.tagName === 'UL') {
      var validNodes = (0, _utils.getValidNodes)(event.target);

      if (validNodes.length > 0 && (0, _keyboard.match)(event, _keyboard.keys.ArrowUp)) {
        nodeToFocus = validNodes[validNodes.length - 1];
      } else if (validNodes.length > 0 && (0, _keyboard.match)(event, _keyboard.keys.ArrowDown)) {
        nodeToFocus = validNodes[0];
      }
    }

    focusNode(nodeToFocus);

    if (rest.onKeyDown) {
      rest.onKeyDown(event);
    }
  }

  function handleClick(e) {
    if (!(0, _utils.clickedElementHasSubnodes)(e) && e.target.tagName !== 'UL') {
      onClose();
    } else {
      e.stopPropagation();
    }
  }

  function handleClickOutside(e) {
    if (!(0, _utils.clickedElementHasSubnodes)(e) && open && canBeClosed) {
      onClose();
    }
  }

  function getCorrectedPosition(assumedDirection) {
    var _rootRef$current2, _rootRef$current2$ele, _getParentMenu, _getParentMenu$getBou, _rootRef$current3;

    var pos = [x, y];

    var _rootRef$current$elem = rootRef === null || rootRef === void 0 ? void 0 : (_rootRef$current2 = rootRef.current) === null || _rootRef$current2 === void 0 ? void 0 : (_rootRef$current2$ele = _rootRef$current2.element) === null || _rootRef$current2$ele === void 0 ? void 0 : _rootRef$current2$ele.getBoundingClientRect(),
        width = _rootRef$current$elem.width,
        height = _rootRef$current$elem.height;

    var _document$body = document.body,
        bodyWidth = _document$body.clientWidth,
        bodyHeight = _document$body.clientHeight;
    var parentWidth = isRootMenu ? 0 : (_getParentMenu = (0, _utils.getParentMenu)(rootRef === null || rootRef === void 0 ? void 0 : (_rootRef$current3 = rootRef.current) === null || _rootRef$current3 === void 0 ? void 0 : _rootRef$current3.element)) === null || _getParentMenu === void 0 ? void 0 : (_getParentMenu$getBou = _getParentMenu.getBoundingClientRect()) === null || _getParentMenu$getBou === void 0 ? void 0 : _getParentMenu$getBou.width;
    var localDirection = assumedDirection;
    var min = [margin, margin];
    var max = [bodyWidth - margin - width, bodyHeight - margin - height]; // in case it is root menu previously had direction -1, check
    // if direction 1 would be possible

    if (isRootMenu && localDirection === -1 && pos[0] < max[0]) {
      localDirection = 1;
    } // make sure menu is visible in y bounds


    if (pos[1] > max[1]) {
      pos[1] = max[1];
    }

    if (pos[1] < min[1]) {
      pos[1] = min[1];
    }

    if (localDirection === 1) {
      // if it won't fit anymore
      if (pos[0] > max[0]) {
        pos[0] = x - width - parentWidth;

        if (pos[0] + width > bodyWidth - margin) {
          pos[0] = max[0];
        }

        localDirection = -1;
      } else if (pos[0] < min[0]) {
        // keep distance to left screen edge
        pos[0] = min[0];
      }
    } else if (localDirection === -1) {
      pos[0] = x - width - parentWidth; // if it should re-reverse

      if (pos[0] < min[0]) {
        pos[0] = x;
        localDirection = 1;
      }
    }

    setDirection(localDirection);
    return [Math.round(pos[0]), Math.round(pos[1])];
  }

  (0, _react.useEffect)(function () {
    setCanBeClosed(false);

    if (open) {
      var localDirection = 1;

      if (isRootMenu) {
        var _rootRef$current4, _rootRef$current4$ele;

        rootRef === null || rootRef === void 0 ? void 0 : (_rootRef$current4 = rootRef.current) === null || _rootRef$current4 === void 0 ? void 0 : (_rootRef$current4$ele = _rootRef$current4.element) === null || _rootRef$current4$ele === void 0 ? void 0 : _rootRef$current4$ele.focus();
      } else {
        var _rootRef$current5;

        var parentMenu = (0, _utils.getParentMenu)(rootRef === null || rootRef === void 0 ? void 0 : (_rootRef$current5 = rootRef.current) === null || _rootRef$current5 === void 0 ? void 0 : _rootRef$current5.element);

        if (parentMenu) {
          localDirection = Number(parentMenu.dataset.direction);
        }
      }

      var correctedPosition = getCorrectedPosition(localDirection);
      setPosition(correctedPosition); // Safari emits the click event when preventDefault was called on
      // the contextmenu event. This is registered by the ClickListener
      // component and would lead to immediate closing when a user is
      // triggering the menu with ctrl+click. To prevent this, we only
      // allow the menu to be closed after the click event was received.
      // Since other browsers don't emit this event, it's also reset with
      // a 50ms delay after mouseup event was called.

      document.addEventListener('mouseup', function () {
        setTimeout(function () {
          setCanBeClosed(true);
        }, 50);
      }, {
        once: true
      });
      document.addEventListener('click', function () {
        setCanBeClosed(true);
      }, {
        once: true
      });
    } else {
      setPosition([0, 0]);
    } // eslint-disable-next-line react-hooks/exhaustive-deps

  }, [open, x, y]);

  var someNodesHaveIcons = _react.default.Children.toArray(children).some(function (node) {
    return node.type === _ContextMenuSelectableItem.default || node.type === _ContextMenuRadioGroup.default;
  });

  var options = _react.default.Children.map(children, function (node) {
    if ( /*#__PURE__*/_react.default.isValidElement(node)) {
      return /*#__PURE__*/_react.default.cloneElement(node, {
        indented: someNodesHaveIcons,
        level: level
      });
    }
  });

  var classes = (0, _classnames2.default)("".concat(prefix, "--context-menu"), (_classnames = {}, _defineProperty(_classnames, "".concat(prefix, "--context-menu--open"), open), _defineProperty(_classnames, "".concat(prefix, "--context-menu--invisible"), open && position[0] === 0 && position[1] === 0), _defineProperty(_classnames, "".concat(prefix, "--context-menu--root"), isRootMenu), _classnames));
  var ulAttributes = {
    className: classes,
    onKeyDown: handleKeyDown,
    onClick: handleClick,
    role: 'menu',
    tabIndex: -1,
    'data-direction': direction,
    'data-level': level,
    style: {
      left: "".concat(position[0], "px"),
      top: "".concat(position[1], "px")
    }
  };
  var childrenToRender = options; // if the only child is a radiogroup, don't render it as radiogroup component, but
  // only the items to prevent duplicate markup

  if (options && options.length === 1 && options[0].type === _ContextMenuRadioGroup.default) {
    var radioGroupProps = options[0].props;
    ulAttributes['aria-label'] = radioGroupProps.label;
    childrenToRender = /*#__PURE__*/_react.default.createElement(_ContextMenuRadioGroupOptions.default, {
      items: radioGroupProps.items,
      initialSelectedItem: radioGroupProps.initialSelectedItem,
      onChange: radioGroupProps.onChange
    });
  } // if the only child is a generic group, don't render it as group component, but
  // only the children to prevent duplicate markup


  if (options && options.length === 1 && options[0].type === _ContextMenuGroup.default) {
    var groupProps = options[0].props;
    ulAttributes['aria-label'] = groupProps.label;
    childrenToRender = _react.default.Children.toArray(options[0].props.children);
  }

  return /*#__PURE__*/_react.default.createElement(_ClickListener.default, {
    onClickOutside: handleClickOutside,
    ref: rootRef
  }, /*#__PURE__*/_react.default.createElement("ul", ulAttributes, childrenToRender));
};

ContextMenu.propTypes = {
  /**
   * Specify the children of the ContextMenu
   */
  children: _propTypes.default.node,

  /**
   * Internal: keeps track of the nesting level of the menu
   */
  level: _propTypes.default.number,

  /**
   * Function called when the menu is closed
   */
  onClose: _propTypes.default.func,

  /**
   * Specify whether the ContextMenu is currently open
   */
  open: _propTypes.default.bool,

  /**
   * Specify the x position where this menu is rendered
   */
  x: _propTypes.default.number,

  /**
   * Specify the y position where this menu is rendered
   */
  y: _propTypes.default.number
};
var _default = ContextMenu;
exports.default = _default;
