// carbon-components-react/lib/components/ContextMenu/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ContextMenuDivider", {
  enumerable: true,
  get: function get() {
    return _ContextMenuDivider.default;
  }
});
Object.defineProperty(exports, "ContextMenuGroup", {
  enumerable: true,
  get: function get() {
    return _ContextMenuGroup.default;
  }
});
Object.defineProperty(exports, "ContextMenuItem", {
  enumerable: true,
  get: function get() {
    return _ContextMenuItem.default;
  }
});
Object.defineProperty(exports, "ContextMenuRadioGroup", {
  enumerable: true,
  get: function get() {
    return _ContextMenuRadioGroup.default;
  }
});
Object.defineProperty(exports, "ContextMenuSelectableItem", {
  enumerable: true,
  get: function get() {
    return _ContextMenuSelectableItem.default;
  }
});
Object.defineProperty(exports, "useContextMenu", {
  enumerable: true,
  get: function get() {
    return _useContextMenu.default;
  }
});
exports.default = void 0;

var _ContextMenu = _interopRequireDefault(require("./ContextMenu"));

var _ContextMenuDivider = _interopRequireDefault(require("./ContextMenuDivider"));

var _ContextMenuGroup = _interopRequireDefault(require("./ContextMenuGroup"));

var _ContextMenuItem = _interopRequireDefault(require("./ContextMenuItem"));

var _ContextMenuRadioGroup = _interopRequireDefault(require("./ContextMenuRadioGroup"));

var _ContextMenuSelectableItem = _interopRequireDefault(require("./ContextMenuSelectableItem"));

var _useContextMenu = _interopRequireDefault(require("./useContextMenu"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Copyright IBM Corp. 2020
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
_ContextMenu.default.ContextMenuDivider = _ContextMenuDivider.default;
_ContextMenu.default.ContextMenuGroup = _ContextMenuGroup.default;
_ContextMenu.default.ContextMenuItem = _ContextMenuItem.default;
_ContextMenu.default.ContextMenuRadioGroup = _ContextMenuRadioGroup.default;
_ContextMenu.default.ContextMenuSelectableItem = _ContextMenuSelectableItem.default;
var _default = _ContextMenu.default;
exports.default = _default;
