// carbon-components-react/lib/components/FormLabel/FormLabel-story.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WithTooltip = exports._Default = exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _FormLabel = _interopRequireDefault(require("./FormLabel"));

var _Tooltip = _interopRequireDefault(require("../Tooltip"));

var _FormLabel2 = _interopRequireDefault(require("./FormLabel.mdx"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
var _default = {
  title: 'Components/FormLabel',
  parameters: {
    component: _FormLabel.default,
    docs: {
      page: _FormLabel2.default
    }
  }
};
exports.default = _default;

var _Default = function _Default() {
  return /*#__PURE__*/_react.default.createElement(_FormLabel.default, null, "Form label");
};

exports._Default = _Default;
_Default.story = {
  name: 'Form Label'
};

var WithTooltip = function WithTooltip() {
  return /*#__PURE__*/_react.default.createElement(_FormLabel.default, null, /*#__PURE__*/_react.default.createElement(_Tooltip.default, {
    triggerText: "Form label"
  }, "This can be used to provide more information about a field."));
};

exports.WithTooltip = WithTooltip;
WithTooltip.story = {
  name: 'Form Label with Tooltip'
};
