// carbon-components-react/lib/components/CopyButton/CopyButton-story.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Playground = exports._Default = exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _addonActions = require("@storybook/addon-actions");

var _addonKnobs = require("@storybook/addon-knobs");

var _CopyButton = _interopRequireDefault(require("../CopyButton"));

var _CopyButton2 = _interopRequireDefault(require("./CopyButton.mdx"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
var props = function props() {
  return {
    feedback: (0, _addonKnobs.text)('The text shown upon clicking (feedback)', 'Copied!'),
    feedbackTimeout: (0, _addonKnobs.number)('How long the text is shown upon clicking (feedbackTimeout)', 3000),
    iconDescription: (0, _addonKnobs.text)('Feedback icon description (iconDescription)', 'Copy to clipboard'),
    onClick: (0, _addonActions.action)('onClick')
  };
};

var _default = {
  title: 'Components/CopyButton',
  decorators: [_addonKnobs.withKnobs],
  parameters: {
    component: _CopyButton.default,
    docs: {
      page: _CopyButton2.default
    }
  }
};
exports.default = _default;

var _Default = function _Default() {
  return /*#__PURE__*/_react.default.createElement(_CopyButton.default, null);
};

exports._Default = _Default;
_Default.story = {
  name: 'Copy Button'
};

var Playground = function Playground() {
  return /*#__PURE__*/_react.default.createElement(_CopyButton.default, props());
};

exports.Playground = Playground;
