// carbon-components-react/lib/components/StructuredList/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StructuredListWrapper = StructuredListWrapper;
exports.StructuredListHead = StructuredListHead;
exports.StructuredListInput = StructuredListInput;
exports.StructuredListBody = StructuredListBody;
exports.StructuredListRow = StructuredListRow;
exports.StructuredListCell = StructuredListCell;
Object.defineProperty(exports, "StructuredListSkeleton", {
  enumerable: true,
  get: function get() {
    return _StructuredList3.default;
  }
});

var _react = _interopRequireDefault(require("react"));

var _StructuredList = require("./next/StructuredList");

var _StructuredList2 = require("./StructuredList");

var _FeatureFlags = require("../FeatureFlags");

var _StructuredList3 = _interopRequireDefault(require("./StructuredList.Skeleton"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
function StructuredListWrapper(props) {
  var enabled = (0, _FeatureFlags.useFeatureFlag)('enable-v11-release');

  if (enabled) {
    return /*#__PURE__*/_react.default.createElement(_StructuredList.StructuredListWrapper, props);
  }

  return /*#__PURE__*/_react.default.createElement(_StructuredList2.StructuredListWrapper, props);
}

function StructuredListHead(props) {
  var enabled = (0, _FeatureFlags.useFeatureFlag)('enable-v11-release');

  if (enabled) {
    return /*#__PURE__*/_react.default.createElement(_StructuredList.StructuredListHead, props);
  }

  return /*#__PURE__*/_react.default.createElement(_StructuredList2.StructuredListHead, props);
}

function StructuredListInput(props) {
  var enabled = (0, _FeatureFlags.useFeatureFlag)('enable-v11-release');

  if (enabled) {
    return /*#__PURE__*/_react.default.createElement(_StructuredList.StructuredListInput, props);
  }

  return /*#__PURE__*/_react.default.createElement(_StructuredList2.StructuredListInput, props);
}

function StructuredListBody(props) {
  var enabled = (0, _FeatureFlags.useFeatureFlag)('enable-v11-release');

  if (enabled) {
    return /*#__PURE__*/_react.default.createElement(_StructuredList.StructuredListBody, props);
  }

  return /*#__PURE__*/_react.default.createElement(_StructuredList2.StructuredListBody, props);
}

function StructuredListRow(props) {
  var enabled = (0, _FeatureFlags.useFeatureFlag)('enable-v11-release');

  if (enabled) {
    return /*#__PURE__*/_react.default.createElement(_StructuredList.StructuredListRow, props);
  }

  return /*#__PURE__*/_react.default.createElement(_StructuredList2.StructuredListRow, props);
}

function StructuredListCell(props) {
  var enabled = (0, _FeatureFlags.useFeatureFlag)('enable-v11-release');

  if (enabled) {
    return /*#__PURE__*/_react.default.createElement(_StructuredList.StructuredListCell, props);
  }

  return /*#__PURE__*/_react.default.createElement(_StructuredList2.StructuredListCell, props);
}
