// carbon-components-react/lib/components/Heading/Heading-story.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.heading = exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _Heading = require("../Heading");

var _Heading2 = _interopRequireDefault(require("./Heading.mdx"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
var _default = {
  title: 'Experimental/unstable_Heading',
  component: _Heading.Heading,
  subcomponents: {
    Section: _Heading.Section
  },
  parameters: {
    docs: {
      page: _Heading2.default
    }
  }
};
exports.default = _default;

var heading = function heading() {
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_Heading.Heading, null, "h1"), /*#__PURE__*/_react.default.createElement(_Heading.Section, null, /*#__PURE__*/_react.default.createElement(_Heading.Heading, null, "h2"), /*#__PURE__*/_react.default.createElement(_Heading.Section, null, /*#__PURE__*/_react.default.createElement(_Heading.Heading, null, "h3"))));
};

exports.heading = heading;
