// carbon-components-react/lib/components/SkeletonPlaceholder/SkeletonPlaceholder-story.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._SkeletonPlaceholder = exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _addonKnobs = require("@storybook/addon-knobs");

var _SkeletonPlaceholder2 = _interopRequireDefault(require("../SkeletonPlaceholder"));

var _SkeletonPlaceholder3 = _interopRequireDefault(require("./SkeletonPlaceholder.mdx"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

/* eslint-disable no-console */
var classNames = {
  'my--skeleton__placeholder--small': 'my--skeleton__placeholder--small',
  'my--skeleton__placeholder--medium': 'my--skeleton__placeholder--medium',
  'my--skeleton__placeholder--large': 'my--skeleton__placeholder--large'
};

var props = function props() {
  return {
    className: (0, _addonKnobs.select)('Classes with different sizes', classNames)
  };
};

var _default = {
  title: 'Components/Skeleton',
  decorators: [_addonKnobs.withKnobs],
  parameters: {
    component: _SkeletonPlaceholder2.default,
    docs: {
      page: _SkeletonPlaceholder3.default
    }
  }
};
exports.default = _default;

var _SkeletonPlaceholder = function _SkeletonPlaceholder() {
  return /*#__PURE__*/_react.default.createElement("div", {
    style: {
      height: '250px',
      width: '250px'
    }
  }, /*#__PURE__*/_react.default.createElement("style", {
    dangerouslySetInnerHTML: {
      __html: "\n      .my--skeleton__placeholder--small {\n        height: 100px;\n        width: 100px;\n      }\n\n      .my--skeleton__placeholder--medium {\n        height: 150px;\n        width: 150px;\n      }\n\n      .my--skeleton__placeholder--large {\n        height: 250px;\n        width: 250px;\n      }\n    "
    }
  }), /*#__PURE__*/_react.default.createElement(_SkeletonPlaceholder2.default, props()));
};

exports._SkeletonPlaceholder = _SkeletonPlaceholder;
